### DISCUSSION TITLE
Can we make freezing possible?

### DISCUSSION THREAD

**Discussion by hmdne** [opal/opal#2294] [Ideas]
It seems like JS has a freezing method with similar semantics...

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze

**Comment by elia** [comment]
I think I tried that in the past, but it's a vague memory, the main issue that can arise and we should verify is if an object has some lazy internal props that it will change even for a frozen object, e.g. if the singleton class is lazily initialized and it's called after a freeze what happens?

**Comment by hmdne** [comment]
(posting my notes from a discussion with @janbiedermann on Slack)

if Object.freeze can be made to work correctly with Opal, we could implement freeze globally

> Any attempt to do so will fail, either silently or by throwing a TypeError exception (most commonly, but not exclusively, when in [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode)).

this doesn't fare well, though
TypeError is quite a wide exception, I thought we could bridge it perhaps
anyway requirement of strict mode is also a bad thing

```js
[user@localhost build]# echo '"use strict"; var a = {}; Object.freeze(a); a.b = 3;' | node
[stdin]:1
"use strict"; var a = {}; Object.freeze(a); a.b = 3;
                                                ^

TypeError: Cannot add property b, object is not extensible
```

hm, Ruby rescue clause is I think done by doing ExceptionType === e
then we can make an ExceptionType a proc, that would try to match an exception for the words in message
and we could force wrap more corelib modules to force "use strict"
the messages that I see that we may handle are:

> Cannot add property b, object is not extensible
> Cannot assign to read only property 'b' of object '#<Object\>'
> Cannot delete property 'b' of #<Object\>

but then quite a lot of wrapping will be involved...
on the other hand, gjs (Gecko) throws different errors:

> can't define property "b": Object is not extensible
> "b" is read-only
> property "b" is non-configurable and can't be deleted

I also tried quickjs which gives:

> object is not extensible
> 'b' is read-only
> could not delete property

also other problem I see is that... freezing is quite an unstable API with many quirks:

```js
Object.isFrozen(1);
// TypeError: 1 is not an object (ES5 code)

Object.isFrozen(1);
// true                          (ES2015 code)
```

**Comment by janbiedermann** [comment]
Object.isFrozen(o) is over 90% slower than tesing for a flag o.$$frozen

**Comment by janbiedermann** [comment]
When freezing Objects the lazy definition of properties used as cache is a problem. So Object.freeze objects and loose the cache/performance or dont Object.freeze them and just rely on the flag.

**Comment by janbiedermann** [comment]
This happens:
```
Uncaught TypeError: Cannot define property $$meta, object is not extensible
    at Function.defineProperty (<anonymous>)
```

**Comment by janbiedermann** [comment]
Yes, we can, in general, with a few peculiarities, WIP here: https://github.com/opal/opal/pull/2444
